packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4108,6 +4107,64 @@ mixin RenderObjectWithChildMixin<ChildType extends RenderObject> on RenderObject
   }
 }
 
+/// A mixin for [RenderObject] subclasses with a layout callback. The mixin
+/// guarantees the layout callback will be called even if this [RenderObject]
+/// skips doing layout, unless the [RenderObject] has never been laid out and
+/// does not have valid [constraints].
+///
+/// A layout callback is a callback that mutates the [RenderObject]'s render
+/// subtree, invoked within an [invokeLayoutCallback] during the [RenderObject]'s
+/// layout process. Sometimes such callbacks involve rebuilding dirty widgets,
+/// especially if the widget configuration depends on the layout process (most
+/// notably, the [LayoutBuilder] widget). Unlike render subtrees, typically all
+/// dirty widgets (even the off-screen ones) in a widget tree must be rebuilt.
+/// This mixin makes sure the layout callback will be invoked (such that dirty
+/// [Element]s are rebuilt) even when an ancestor [RenderObject] chose to skip
+/// laying out this [RenderObject].
+///
+/// Subclasses must not invoke the layout callback directly. Instead, call
+/// `super.performLayout` in the [performLayout] implementation.
+///
+/// See also:
+///
+///  * [LayoutBuilder] and [SliverLayoutBuilder], which use the mixin.
+mixin RenderObjectWithLayoutCallbackMixin on RenderObject {
+  // The initial value of this flag must be set to true to prevent the layout
+  // callback from being scheduled when the subtree has never been laid out
+  // (in which case the `constraints` is unknown).
+  bool _needsRebuild = true;
+
+  @mustCallSuper
+  @override
+  void performLayout() {
+    invokeLayoutCallback(runLayoutCallback);
+    _needsRebuild = false;
+  }
+
+  /// The layout callback to be invoked during [performLayout].
+  ///
+  /// This method should not be invoked directly. Instead, call
+  /// `super.performLayout` in the [performLayout] implementation. This
+  /// implementation will be invoked within [invokeLayoutCallback].
+  @visibleForOverriding
+  void runLayoutCallback(covariant Constraints constraints);
+
+  /// Informs the framework that the layout callback has been updated and must
+  /// be invoked again.
```

Comment:
Nit: Consider explaining the difference between this and markNeedsLayout here. Also maybe add a "see also" for this method in the docs of markNeedsLayout?

Reply:
I think in this mixin, it's easier to think just in terms of layout callback and ignore `markNeedsLayout`? I.e. treat them as orthogonal concepts like markNeedsLayout vs. markNeedsPaint, if the layout callback is changed, call scheduleLayoutCallback, and if the layout needs changing, call markNeedsLayout.

Reply:
That's true, looks good as is :+1: . I was worrying that someone will see both methods in their IDE autocomplete or something, but that is unlikely.